File: C++/Battleships in a Board.cpp
/*O(1) extra memory*/
class Solution {
    private:
        int row_size = 0;
        int column_size = 0;
        int count_Battleships = 0;
        
public:
    int countBattleships(vector<vector<char>>& board) {
        row_size = board.size();
        column_size = board[0].size();
        for(int i=0;i<row_size;i++)
        {
            for(int j=0;j<column_size;j++)
            {
                  if (board[i][j] == 'X' && !(i > 0 && board[i - 1][j] == 'X' || j > 0 && board[i][j - 1] == 'X')) 
                  
                  {
                    
                     count_Battleships += 1;
                
                    }
            }
        }
        
        return count_Battleships;
    }
};

File: Meta/Practice/Python/Battleships in a board.py
'''
Given an m x n matrix board where each cell is a battleship 'X' or empty '.', return the number of the battleships on board.

Battleships can only be placed horizontally or vertically on board.
In other words, they can only be made of the shape 1 x k (1 row, k columns) or k x 1 (k rows, 1 column), where k can be of any size. 
At least one horizontal or vertical cell separates between two battleships (i.e., there are no adjacent battleships).

Input: board = [["X",".",".","X"],[".",".",".","X"],[".",".",".","X"]]
Output: 2

Constraints:

m == board.length
n == board[i].length
1 <= m, n <= 200
board[i][j] is either '.' or 'X'.

Follow up: Could you do it in one-pass, using only O(1) extra memory and without modifying the values board?
'''

# Using only O(1) extra memory and without modifying the value on board
'''
Comlexity
Tc:- O(m*n)
Sc:- O(1)
'''
class Solution:
    def countBattleships(self, board: List[List[str]]) -> int:
        if not board:
            return 0
        
        output =0
        
        for i in range(len(board)):
            for j in range(len(board[0])):
                if board[i][j]=='.':
                    continue
                if i > 0 and board[i-1][j] == 'X':
                    continue
                if j > 0 and board[i][j-1] == 'X':
                    continue
                output +=1
        return output
